scene/gui/color_picker.cpp

```diff
@@ -353,27 +320,94 @@ void ColorPicker::_value_changed(double) {
 		return;
 	}
 
-	if (hsv_mode_enabled) {
-		h = scroll[0]->get_value() / 360.0;
-		s = scroll[1]->get_value() / 100.0;
-		v = scroll[2]->get_value() / 100.0;
-		if (picker_type == SHAPE_OKHSL_CIRCLE) {
-			color.set_ok_hsl(h, s, v, Math::round(scroll[3]->get_value() / 255.0));
-		} else {
-			color.set_hsv(h, s, v, Math::round(scroll[3]->get_value() / 255.0));
-		}
+	color = selected_mode->get_color();
 
+	if (mode_option_button->get_selected() == MODE_HSV || mode_option_button->get_selected() == MODE_OKHSL) {
+		h = sliders[0]->get_value() / 360.0;
+		s = sliders[1]->get_value() / 100.0;
+		v = sliders[2]->get_value() / 100.0;
 		last_color = color;
-	} else {
-		for (int i = 0; i < 4; i++) {
-			color.components[i] = scroll[i]->get_value() / (raw_mode_enabled ? 1.0 : 255.0);
-		}
 	}
 
 	_set_pick_color(color, false);
 	emit_signal(SNAME("color_changed"), color);
 }
 
+void ColorPicker::add_mode(ColorMode *p_mode) {
+	modes.push_back(p_mode);
+	mode_option_button->add_item(RTR(p_mode->get_name()));
+}
+
+void ColorPicker::create_slider(GridContainer *gc, int idx) {
+	Label *l = memnew(Label());
+	l->set_custom_minimum_size(Size2(get_theme_constant(SNAME("label_width")), 0));
+	l->set_v_size_flags(SIZE_SHRINK_CENTER);
+	gc->add_child(l);
+
+	HSlider *s = memnew(HSlider);
+	s->set_v_size_flags(SIZE_SHRINK_CENTER);
+	s->set_focus_mode(FOCUS_NONE);
+	gc->add_child(s);
+
+	SpinBox *v = memnew(SpinBox);
+	s->share(v);
+	gc->add_child(v);
+	v->get_line_edit()->connect("focus_entered", callable_mp(this, &ColorPicker::_focus_enter));
+	v->get_line_edit()->connect("focus_exited", callable_mp(this, &ColorPicker::_focus_exit));
+
+	s->set_min(0);
+	s->set_page(0);
+	s->set_h_size_flags(SIZE_EXPAND_FILL);
+
+	s->connect("value_changed", callable_mp(this, &ColorPicker::_value_changed));
+	s->connect("draw", callable_mp(this, &ColorPicker::_slider_draw), make_binds(idx));
+
+	if (idx < SLIDER_COUNT) {
+		sliders[idx] = s;
+		values[idx] = v;
+		labels[idx] = l;
+	} else {
+		alpha_slider = s;
+		alpha_value = v;
+		alpha_label = l;
+	}
+}
+
+Vector<HSlider *> ColorPicker::get_sliders() {
+	Vector<HSlider *> hs;
+	for (int i = 0; i < SLIDER_COUNT; i++) {
+		hs.push_back(sliders[i]);
+	}
+	hs.push_back(alpha_slider);
+	return hs;
+}
+
+Vector<float> ColorPicker::get_active_slider_values() {
+	Vector<float> values;
+	for (int i = 0; i < current_slider_count; i++) {
+		values.push_back(sliders[i]->get_value());
+	}
+	values.push_back(alpha_slider->get_value());
+	return values;
+}
```

Comment:
Here is similar to the above, but you are actually using all of the values after fetching them, so it's probably ok to keep it like that.